File: apps/backend-python/src/housing_platform/notifications/service.py
import uuid

from sqlalchemy.orm import Session

from housing_platform.auth.errors import NotFoundError
from housing_platform.auth.models import AuthUser
from housing_platform.notifications.mappers import map_notification
from housing_platform.notifications.repository import NotificationRepository
from housing_platform.notifications.schemas import (
    MarkAllNotificationsReadResult,
    Notification,
    UnreadNotificationCount,
)


class NotificationService:
    def __init__(
        self,
        db: Session,
        repository: NotificationRepository | None = None,
    ) -> None:
        self._db = db
        self._repo = repository or NotificationRepository(db)

    def list_notifications(self, user: AuthUser) -> list[Notification]:
        rows = self._repo.list_for_user(user.id)
        return [map_notification(row) for row in rows]

    def get_unread_count(self, user: AuthUser) -> UnreadNotificationCount:
        return UnreadNotificationCount(count=self._repo.count_unread(user.id))

    def mark_read(self, user: AuthUser, notification_id: uuid.UUID) -> Notification:
        notification = self._repo.mark_read(user.id, notification_id)
        if notification is None:
            raise NotFoundError("Notification not found")

        self._db.commit()
        self._db.refresh(notification)
        return map_notification(notification)

    def mark_all_read(self, user: AuthUser) -> MarkAllNotificationsReadResult:
        updated_count = self._repo.mark_all_read(user.id)
        self._db.commit()
        return MarkAllNotificationsReadResult(updated_count=updated_count)

File: apps/backend-go/internal/notifications/service.go
package notifications

import (
	"context"

	"github.com/jackc/pgx/v5"

	apierrors "github.com/housing-platform/backend-go/internal/api/errors"
	"github.com/housing-platform/backend-go/internal/auth"
	"github.com/housing-platform/backend-go/internal/shared"
)

type Service struct {
	repo *Repository
}

func NewService(repo *Repository) *Service {
	return &Service{repo: repo}
}

func (s *Service) ListNotifications(ctx context.Context, user *auth.User) ([]Notification, error) {
	userID, err := shared.ParseUUID(user.ID)
	if err != nil {
		return nil, err
	}

	var items []Notification
	err = shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		rows, err := s.repo.ListForUser(ctx, tx, userID)
		if err != nil {
			return err
		}
		items = make([]Notification, 0, len(rows))
		for _, row := range rows {
			items = append(items, mapNotification(row))
		}
		return nil
	})
	return items, err
}

func (s *Service) GetUnreadCount(ctx context.Context, user *auth.User) (UnreadNotificationCount, error) {
	userID, err := shared.ParseUUID(user.ID)
	if err != nil {
		return UnreadNotificationCount{}, err
	}

	var count int
	err = shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		var err error
		count, err = s.repo.CountUnread(ctx, tx, userID)
		return err
	})
	return UnreadNotificationCount{Count: count}, err
}

func (s *Service) MarkRead(ctx context.Context, user *auth.User, notificationID string) (Notification, error) {
	userID, err := shared.ParseUUID(user.ID)
	if err != nil {
		panic(apierrors.BadRequest("Invalid user ID"))
	}
	notifUUID, err := shared.ParseUUID(notificationID)
	if err != nil {
		panic(apierrors.BadRequest("Invalid notification ID"))
	}

	var result Notification
	err = shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		row, err := s.repo.MarkRead(ctx, tx, userID, notifUUID)
		if err != nil {
			return err
		}
		if row == nil {
			panic(apierrors.NotFound("Notification not found"))
		}
		result = mapNotification(*row)
		return nil
	})
	return result, err
}

func (s *Service) MarkAllRead(ctx context.Context, user *auth.User) (MarkAllNotificationsReadResult, error) {
	userID, err := shared.ParseUUID(user.ID)
	if err != nil {
		return MarkAllNotificationsReadResult{}, err
	}

	var updated int
	err = shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		var err error
		updated, err = s.repo.MarkAllRead(ctx, tx, userID)
		return err
	})
	return MarkAllNotificationsReadResult{UpdatedCount: updated}, err
}
